### PR TITLE
add flag for turn shutdown complete. At turnConnectionShutdown, run c…

### LINKED ISSUE #649: [BUG] Sudden CPU increase after  connectionListenerReceiveDataRoutine(): recvfrom() failed with errno Connection reset by peer for socket 15
We have observed a correlation between this event and a sudden CPU increase after 30 seconds into streaming.
This will make our device CPU usage to increase from 20% to 60% and stays at 60%. If this event does not occur by luck, then the CPU stays at 20% utilization. Regardless of what code path, the video stream continues to play.

Sample events leading up to sudden CPU increase.
```
2020-07-23 06:23:10 VERBOSE iceAgentSendKeepAliveTimerCallback(): send keep alive
2020-07-23 06:23:10 VERBOSE iceAgentSendKeepAliveTimerCallback(): send keep alive
2020-07-23 06:23:10 VERBOSE iceAgentSendKeepAliveTimerCallback(): send keep alive
2020-07-23 06:23:10 DEBUG   socketConnectionClosed(): Close socket 15
2020-07-23 06:23:10 DEBUG   connectionListenerReceiveDataRoutine(): recvfrom() failed with errno Connection reset by peer for socket 15
2020-07-23 06:23:10 VERBOSE signalingClientGetCurrentState(): Signaling Client Get Current State
```

### PR DESCRIPTION
**Pull request by chehefen**
…leanup regardless of having allocation or not to clean up the socket. Add unit test covering turn socket clean up

*Issue #, if available:*
fix #649 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/Ice/ConnectionListener.c [hunk 1]
```diff
@@ -129,6 +131,19 @@ STATUS connectionListenerRemoveConnection(PConnectionListener pConnectionListene
...
     MUTEX_LOCK(pConnectionListener->lock);
     locked = TRUE;
+
+    CHK_STATUS(doubleListGetHeadNode(pConnectionListener->connectionList, &pCurNode));
+    while (!hasConnection && pCurNode != NULL) {
+        pCurrSocketConnection = (PSocketConnection) pCurNode->data;
+        pCurNode = pCurNode->pNext;
+        if (pCurrSocketConnection == pSocketConnection) {
+            hasConnection = TRUE;
+        }
```
Comment (MushMal) [inline src/source/Ice/ConnectionListener.c:141]: Nitpick. hasConnection = (pCurrSocketConnection == pSocketConnection);

File: src/source/Ice/TurnConnection.c [hunk 2]
```diff
@@ -211,6 +218,8 @@ STATUS turnConnectionHandleStun(PTurnConnection pTurnConnection, PBYTE pBuffer,
 
     switch (stunPacketType) {
         case STUN_PACKET_TYPE_ALLOCATE_SUCCESS_RESPONSE:
+            /* If shutdown has been initiated, ignore the alloaction response */
```
Comment (MushMal) [inline src/source/Ice/TurnConnection.c:221]: *allocation

File: src/source/Ice/TurnConnection.c [hunk 3]
```diff
@@ -1048,10 +1057,9 @@ STATUS turnConnectionStepState(PTurnConnection pTurnConnection)
                 }
 
                 CHK_STATUS(turnConnectionFreePreAllocatedPackets(pTurnConnection));
-
+                CHK_STATUS(socketConnectionClosed(pTurnConnection->pControlChannel));
```
Comment (MushMal) [inline src/source/Ice/TurnConnection.c:1060]: I am wondering if you still need to set the shutdownComplete atomic in case of failure anyway?

Comment (chehefen) [reply src/source/Ice/TurnConnection.c:1060]: I added logic that when failure happens, turn will go to cleanup state, then go to failed state. Also added unit test for that.

**Comment by chehefen** [comment]
> This looks good to me overall. Could you also add a unit test that closes one of the sockets from the connection list, and verify that it's deleted from the list properly by waiting on `pConnectionListener->removeConnectionComplete` condition variable?
In the two new unit tests I added, at the end I added check for turnSocketConnection to not exist in connectionListener's connection list.

**Pull request merged**